config/src/keys.rs

```diff
@@ -99,22 +45,66 @@ where
 impl<T> KeyPair<T>
 where
     T: PrivateKey + Serialize + ValidKeyStringExt,
-    T::PublicKeyMaterial: DeserializeOwned + 'static + Serialize + ValidKeyStringExt,
 {
+    /// This transforms a private key into a keypair data structure.
     pub fn load(private_key: T) -> Self {
         let public_key = private_key.public_key();
         Self {
-            private_key: PrivateKeyContainer::Present(private_key),
+            private_key: Some(private_key),
             public_key,
         }
     }
 
-    pub fn public(&self) -> &T::PublicKeyMaterial {
-        &self.public_key
-    }
-
-    /// Beware, this destroys the private key from this NodeConfig
+    /// Takes the key from the data structure, calling this function a second time will return None.
     pub fn take_private(&mut self) -> Option<T> {
         self.private_key.take()
     }
+
+    /// Returns the public key part. This always work, even after the private key was taken.
+    pub fn public_key(&self) -> T::PublicKeyMaterial {
+        self.public_key.clone()
```

Comment:
Why aren't we returning the reference? It seems just as much callers want a copy as they want the reference, prefer reference.

Reply:
a public key is 32-byte which is cheap enough to copy that it sounds like we should just handle these like values, especially since we only use this at initialization (I originally had the Copy trait on x25519 public keys). What do you think? Don't have a strong opinion but I was wondering what is more "idiomatic".

Reply:
Then derive copy but please don't spew clone all over the place :P

Reply:
I see what you're saying: better to return a borrow which will use copy if the receiving type asks for it. Is that right?

Reply:
OK this is tripping me, so I looked a bit further and it looks like when we use a reference of public_key (for example `AuthenticationKey::try_from`) we end up copying the value anyway.

Interestingly, the rust std lib has [more to say about this](https://doc.rust-lang.org/std/marker/trait.Copy.html#when-should-my-type-be-copy)!

> Generally speaking, if your type can implement Copy, it should. Keep in mind, though, that implementing Copy is part of the public API of your type. If the type might become non-Copy in the future, it could be prudent to omit the Copy implementation now, to avoid a breaking API change.

So I really think this is the perfect use-case for deriving `Copy`.